nav2_smac_planner/src/smac_planner_2d.cpp

```diff
@@ -285,6 +293,100 @@ nav_msgs::msg::Path SmacPlanner2D::createPlan(
   return plan;
 }
 
+void SmacPlanner2D::on_parameter_event_callback(
+  const rcl_interfaces::msg::ParameterEvent::SharedPtr event)
+{
+  bool reinit_a_star = false;
+  bool reinit_downsampler = false;
+
+  for (auto & changed_parameter : event->changed_parameters) {
+    const auto & type = changed_parameter.value.type;
+    const auto & name = changed_parameter.name;
+    const auto & value = changed_parameter.value;
+
+    if (type == ParameterType::PARAMETER_DOUBLE) {
+      if (name == _name + ".tolerance") {
+        _tolerance = static_cast<float>(value.double_value);
+      } else if (name == _name + ".cost_travel_multiplier") {
+        reinit_a_star = true;
+        _search_info.cost_penalty = value.double_value;
+      } else if (name == _name + ".max_planning_time") {
+        _max_planning_time = value.double_value;
+      }
+    } else if (type == ParameterType::PARAMETER_BOOL) {
+      if (name == _name + ".downsample_costmap") {
+        reinit_downsampler = true;
+        _downsample_costmap = value.bool_value;
+      } else if (name == _name + ".allow_unknown") {
+        reinit_a_star = true;
+        _allow_unknown = value.bool_value;
+      }
+    } else if (type == ParameterType::PARAMETER_INTEGER) {
+      if (name == _name + ".downsampling_factor") {
+        reinit_downsampler = true;
+        _downsampling_factor = value.integer_value;
+      } else if (name == _name + ".max_iterations") {
+        reinit_a_star = true;
+        _max_iterations = value.integer_value;
+        if (_max_iterations <= 0) {
+          RCLCPP_INFO(
+            _logger, "maximum iteration selected as <= 0, "
+            "disabling maximum iterations.");
+          _max_iterations = std::numeric_limits<int>::max();
+        }
+      } else if (name == _name + ".max_on_approach_iterations") {
+        reinit_a_star = true;
+        _max_on_approach_iterations = value.integer_value;
+        if (_max_on_approach_iterations <= 0) {
+          RCLCPP_INFO(
+            _logger, "On approach iteration selected as <= 0, "
+            "disabling tolerance and on approach iterations.");
+          _max_on_approach_iterations = std::numeric_limits<int>::max();
+        }
+      }
+    } else if (type == ParameterType::PARAMETER_STRING) {
+      if (name == _name + ".motion_model_for_search") {
+        reinit_a_star = true;
+        _motion_model = fromString(value.string_value);
+        if (_motion_model == MotionModel::UNKNOWN) {
+          RCLCPP_WARN(
+            _logger,
+            "Unable to get MotionModel search type. Given '%s', "
+            "valid options are MOORE, VON_NEUMANN, DUBIN, REEDS_SHEPP.",
+            _motion_model_for_search.c_str());
+        }
+      }
+    }
+  }
+
+  // Re-init if needed with mutex lock (to avoid re-init while creating a plan)
+  if (reinit_a_star || reinit_downsampler) {
+    std::lock_guard<std::mutex> lock_reinit(_mutex);
```

Comment:
Last thing I see, since you are changing a couple of values that can actually change live, this mutex lock needs to happen either earlier in the function or in those specific `if` statements above. As we noted in the thread above, there are 2 (I believe) params that can be changed live and used without reinitalization. So its possible this function could be in the middle of execution to change that value while trying to be read for use.

Reply:
I believe the two parameters that can be changed without re-init are `_tolerance` and `_max_planning_time`, all the others are needed either for `_a_star` or `_costmap_downsampler`. Hence by putting the lock there, I use it only if a re-init is needed.

Reply:
huh? If you change `_tolerance` and `_max_planning_time` you don't need to reinit, yes. But since those are *independent* of the A* / costmap downsampler, they can be changed in this callback at the same time that a planning request is processing. So it could be trying to access that memory at the same time its being overwritten. So we need a lock in the `if` statement where you set those 2 variables in particular since they are not constrained by the A* / downsampler objects that are locked already.

Does that make sense?

- Planner is called at the same time as a reconfigure service
- If any others are changed, doesn't matter since the planning request will finish before giving up the mutex for the reinitialization to happen
- If this is changed, it can be corrupt a `get` while this is `setting` them, since used live in the code and not just within the A*/downsampler objects.

Reply:
All right, I completely overlooked this potential race condition.
Basically all params needs protection so I placed the lock at the top of the param change callback